File: infrastructure/parsers/portal_page_parser.py
from __future__ import annotations

import re
from html import unescape as html_unescape
from html.parser import HTMLParser
from urllib.parse import parse_qs, urljoin, urlsplit

from domain.models.traffic import PortalHiddenFields
from infrastructure.network.models import YiiLoginFormData


class _HiddenInputParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self.values: dict[str, str] = {
            "ac_id": "",
            "user_ip": "",
            "nas_ip": "",
            "user_mac": "",
        }

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        if tag.lower() != "input":
            return
        attr_map = {key.lower(): (value or "") for key, value in attrs}
        name = attr_map.get("name", "")
        if name in self.values:
            self.values[name] = attr_map.get("value", "")


class _YiiLoginFormParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self._form_depth = 0
        self._inside_login_form = False
        self.action_url = "/login"
        self.csrf_name = ""
        self.csrf_value = ""
        self.captcha_url = ""

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attr_map = {key.lower(): (value or "") for key, value in attrs}
        lower_tag = tag.lower()

        if lower_tag == "form":
            self._form_depth += 1
            form_id = attr_map.get("id", "")
            action = attr_map.get("action", "")
            if form_id == "login-form" or "/login" in action:
                self._inside_login_form = True
                self.action_url = action or "/login"
            return

        if lower_tag == "input" and self._inside_login_form:
            name = attr_map.get("name", "")
            if name.startswith("_csrf"):
                self.csrf_name = name
                self.csrf_value = attr_map.get("value", "")
            return

        if lower_tag == "img":
            element_id = attr_map.get("id", "")
            src = attr_map.get("src", "")
            if element_id == "loginform-verifycode-image" or "captcha" in src:
                self.captcha_url = src

    def handle_endtag(self, tag: str) -> None:
        if tag.lower() != "form":
            return
        if self._form_depth > 0:
            self._form_depth -= 1
        if self._inside_login_form and self._form_depth == 0:
            self._inside_login_form = False


def is_yii_login_page(html: str) -> bool:
    lower_html = html.lower()
    return (
        "loginform[username]" in lower_html
        and "loginform[password]" in lower_html
        and "loginform[verifycode]" in lower_html
    )


def is_traffic_home_page(html: str) -> bool:
    return "产品名称" in html and "已用流量" in html and "产品余额" in html


def parse_hidden_fields(html: str, page_url: str) -> PortalHiddenFields:
    parser = _HiddenInputParser()
    parser.feed(html)
    parser.close()

    values = parser.values
    if not values["ac_id"]:
        query = parse_qs(urlsplit(page_url).query, keep_blank_values=True)
        values["ac_id"] = query.get("ac_id", [""])[0]

    return PortalHiddenFields(
        ac_id=values["ac_id"],
        user_ip=values["user_ip"],
        nas_ip=values["nas_ip"],
        user_mac=values["user_mac"],
    )


def parse_yii_login_form(html: str, page_url: str) -> YiiLoginFormData:
    parser = _YiiLoginFormParser()
    parser.feed(html)
    parser.close()

    if not parser.csrf_name or not parser.csrf_value:
        raise RuntimeError("登录页缺少 CSRF 字段，无法提交表单")

    captcha_src = parser.captcha_url or "/site/captcha?refresh=1"
    return YiiLoginFormData(
        csrf_name=parser.csrf_name,
        csrf_value=parser.csrf_value,
        captcha_url=urljoin(page_url, captcha_src),
        action_url=urljoin(page_url, parser.action_url or "/login"),
    )


def extract_yii_error_message(html: str) -> str:
    text = re.sub(r"<[^>]+>", " ", html)
    text = re.sub(r"\s+", " ", text).strip()
    for key in [
        "验证码不正确",
        "用户名不能为空",
        "密码不能为空",
        "用户名或密码错误",
        "请修复以下错误",
    ]:
        if key in text:
            return key
    return text[:80] if text else "登录失败"


def normalize_captcha_code(raw_code: str | None) -> str:
    raw_text = raw_code or ""
    normalized = re.sub(r"[^0-9A-Za-z]", "", raw_text.strip())
    return normalized[:4]


def extract_meta_content(html: str, meta_name: str) -> str:
    patterns = [
        rf'<meta[^>]*name=["\']{re.escape(meta_name)}["\'][^>]*content=["\']([^"\']+)["\']',
        rf'<meta[^>]*content=["\']([^"\']+)["\'][^>]*name=["\']{re.escape(meta_name)}["\']',
    ]
    for pattern in patterns:
        match = re.search(pattern, html or "", flags=re.IGNORECASE)
        if match:
            return html_unescape(match.group(1)).strip()
    return ""

File: src-tauri/src/infrastructure/parsers/portal_page_parser.rs
use scraper::{Html, Selector};
use url::Url;

use crate::application::error::{AppError, AppResult};
use crate::domain::models::PortalHiddenFields;
use crate::infrastructure::network::models::YiiLoginFormData;

pub fn is_yii_login_page(html: &str) -> bool {
    let lower = html.to_lowercase();
    lower.contains("loginform[username]")
        && lower.contains("loginform[password]")
        && lower.contains("loginform[verifycode]")
}

pub fn is_traffic_home_page(html: &str) -> bool {
    html.contains("产品名称") && html.contains("已用流量") && html.contains("产品余额")
}

pub fn parse_hidden_fields(html: &str, page_url: &str) -> PortalHiddenFields {
    let doc = Html::parse_document(html);
    let selector = Selector::parse("input").expect("valid selector");
    let mut fields = PortalHiddenFields::default();
    for element in doc.select(&selector) {
        let name = element.value().attr("name").unwrap_or_default();
        let value = element
            .value()
            .attr("value")
            .unwrap_or_default()
            .to_string();
        match name {
            "ac_id" => fields.ac_id = value,
            "user_ip" => fields.user_ip = value,
            "nas_ip" => fields.nas_ip = value,
            "user_mac" => fields.user_mac = value,
            _ => {}
        }
    }
    if fields.ac_id.is_empty() {
        if let Ok(url) = Url::parse(page_url) {
            if let Some((_, value)) = url.query_pairs().find(|(key, _)| key == "ac_id") {
                fields.ac_id = value.to_string();
            }
        }
    }
    fields
}

pub fn parse_yii_login_form(html: &str, page_url: &str) -> AppResult<YiiLoginFormData> {
    let doc = Html::parse_document(html);
    let input_selector = Selector::parse("input").expect("valid selector");
    let form_selector = Selector::parse("form").expect("valid selector");
    let img_selector = Selector::parse("img").expect("valid selector");

    let mut csrf_name = String::new();
    let mut csrf_value = String::new();
    for input in doc.select(&input_selector) {
        let name = input.value().attr("name").unwrap_or_default();
        if name.starts_with("_csrf") {
            csrf_name = name.to_string();
            csrf_value = input.value().attr("value").unwrap_or_default().to_string();
            break;
        }
    }
    if csrf_name.is_empty() || csrf_value.is_empty() {
        return Err(AppError::Network(
            "登录页缺少 CSRF 字段，无法提交表单".to_string(),
        ));
    }

    let mut action = "/login".to_string();
    for form in doc.select(&form_selector) {
        let id = form.value().attr("id").unwrap_or_default();
        let candidate = form.value().attr("action").unwrap_or_default();
        if id == "login-form" || candidate.contains("/login") {
            action = if candidate.is_empty() {
                "/login".to_string()
            } else {
                candidate.to_string()
            };
            break;
        }
    }

    let mut captcha = "/site/captcha?refresh=1".to_string();
    for img in doc.select(&img_selector) {
        let id = img.value().attr("id").unwrap_or_default();
        let src = img.value().attr("src").unwrap_or_default();
        if id == "loginform-verifycode-image" || src.contains("captcha") {
            captcha = src.to_string();
            break;
        }
    }

    Ok(YiiLoginFormData {
        csrf_name,
        csrf_value,
        captcha_url: join_url(page_url, &captcha),
        action_url: join_url(page_url, &action),
    })
}

pub fn extract_yii_error_message(html: &str) -> String {
    let re = regex::Regex::new(r"<[^>]+>").expect("valid regex");
    let text = re.replace_all(html, " ");
    let text = text.split_whitespace().collect::<Vec<_>>().join(" ");
    for key in [
        "验证码不正确",
        "用户名不能为空",
        "密码不能为空",
        "用户名或密码错误",
        "请修复以下错误",
    ] {
        if text.contains(key) {
            return key.to_string();
        }
    }
    if text.is_empty() {
        "登录失败".to_string()
    } else {
        text.chars().take(80).collect()
    }
}

pub fn normalize_captcha_code(raw_code: &str) -> String {
    raw_code
        .chars()
        .filter(|ch| ch.is_ascii_alphanumeric())
        .take(4)
        .collect()
}

pub fn extract_meta_content(html: &str, meta_name: &str) -> String {
    let doc = Html::parse_document(html);
    let selector = Selector::parse("meta").expect("valid selector");
    for meta in doc.select(&selector) {
        if meta.value().attr("name") == Some(meta_name) {
            return meta
                .value()
                .attr("content")
                .unwrap_or_default()
                .trim()
                .to_string();
        }
    }
    String::new()
}

pub fn join_url(base_url: &str, path: &str) -> String {
    Url::parse(base_url)
        .and_then(|base| base.join(path))
        .map(|url| url.to_string())
        .unwrap_or_else(|_| path.to_string())
}
